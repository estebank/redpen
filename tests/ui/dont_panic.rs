struct S;

impl S {
    fn foo(&self) {
        panic!("");
    }
}

trait A {
    fn from_trait(&self) {
        panic!();
    }
    fn from_a_impl(&self);
}
trait B {
    fn from_impl(&self);
}

impl A for S {
    fn from_a_impl(&self) {
        panic!();
    }
}
impl B for S {
    fn from_impl(&self) {
        panic!();
    }
}

#[redpen::wont_panic]
fn might_panic() {
    panic!()
}

#[deny(redpen::dont_panic)]
fn impl_fn(x: impl A) {
    x.from_trait();
    x.from_a_impl();
}
#[deny(redpen::dont_panic)]
fn trait_object(x: Box<dyn A>) {
    x.from_trait();
    x.from_a_impl();
}

#[deny(redpen::dont_panic)]
fn foo() {
    panic!("shouldn't happen!");
    S.foo();
    S.from_trait();
    S.from_a_impl();
    S.from_impl();
    #[allow(redpen::dont_panic)]
    (|| {
        panic!();
    })();
    #[allow(redpen::dont_panic)]
    let x =  || panic!();
    x();
    let mut x = vec![];
    x.push("");
    let _ = x[1];
    might_panic();
}

#[deny(redpen::dont_panic)]
fn bar() {
    foo();
}

#[deny(redpen::dont_panic)]
fn baz() {
}

fn main() {
    foo();
    bar();
    baz();
}