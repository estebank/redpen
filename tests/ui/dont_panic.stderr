error: `impl_fn` can panic
  --> $DIR/dont_panic.rs:36:1
   |
35 | #[redpen::dont_panic]
   | --------------------- the function can't panic due to this annotation
36 | fn impl_fn(x: impl A) {
   | ^^^^^^^^^^^^^^^^^^^^^ this function can panic
37 |     x.from_trait();
   |       ---------- panic can occur here
38 |     x.from_a_impl();
   |       ----------- panic can occur here
   |
   = note: `#[deny(redpen::dont_panic)]` on by default

error: `trait_object` can panic
  --> $DIR/dont_panic.rs:41:1
   |
40 | #[redpen::dont_panic]
   | --------------------- the function can't panic due to this annotation
41 | fn trait_object(x: Box<dyn A>) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this function can panic
42 |     x.from_trait();
   |       ---------- panic can occur here
43 |     x.from_a_impl();
   |       ----------- panic can occur here

error: `foo` can panic
  --> $DIR/dont_panic.rs:47:1
   |
46 |   #[redpen::dont_panic]
   |   --------------------- the function can't panic due to this annotation
47 |   fn foo() {
   |   ^^^^^^^^ this function can panic
48 |       panic!("shouldn't happen!");
   |       --------------------------- panic can occur here
49 |       S.foo();
   |         --- panic can occur here
50 |       S.from_trait();
   |         ---------- panic can occur here
51 |       S.from_a_impl();
   |         ----------- panic can occur here
52 |       S.from_impl();
   |         --------- panic can occur here
53 | /     (|| {
54 | |         panic!();
55 | |     })();
   | |______- panic can occur here
56 |       let x =  || panic!();
57 |       x();
   |       - panic can occur here
...
60 |       let _ = x[1];
   |               ---- panic can occur here

error: `bar` can panic
  --> $DIR/dont_panic.rs:65:1
   |
64 | #[redpen::dont_panic]
   | --------------------- the function can't panic due to this annotation
65 | fn bar() {
   | ^^^^^^^^ this function can panic
66 |     foo();
   |     --- panic can occur here

error: aborting due to 4 previous errors

